snuba/datasets/generic_metrics_processor.py

```diff
@@ -0,0 +1,116 @@
+import logging
+import zlib
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Mapping, MutableMapping, Optional
+
+from snuba.consumers.types import KafkaMessageMetadata
+from snuba.datasets.events_format import EventTooOld, enforce_retention
+from snuba.datasets.metrics_messages import is_set_message, values_for_set_message
+from snuba.processor import (
+    InsertBatch,
+    MessageProcessor,
+    ProcessedMessage,
+    _ensure_valid_date,
+)
+
+logger = logging.getLogger(__name__)
+
+# These are the hardcoded values from the materialized view
+GRANULARITY_ONE_MINUTE = 1
+GRANULARITY_ONE_HOUR = 2
+GRANULARITY_ONE_DAY = 3
+
+
+class GenericMetricsBucketProcessor(MessageProcessor, ABC):
+    @abstractmethod
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    def _hash_timeseries_id(self, message: Mapping[str, Any]) -> int:
+        """
+        _hash_timeseries_id should return a UInt32 whose distribution should shard
+        as evenly as possible while ensuring that an average query will not have to
+        cross shards to read a results (so for the same org, project, metric, and tags
+        ClickHouse should not have to aggregate results from multiple nodes).
+        """
+        use_case_id: str = message.get("use_case_id") or ""
+        org_id: str = message["org_id"]
+        project_id: str = message["project_id"]
+        metric_id: str = message["metric_id"]
+        tag_keys_comma_sep: str = ",".join(message["tags"].keys())
+
+        return zlib.adler32(
```

Comment:
I think I just picked it because of this sentence in the [docs](https://docs.python.org/3/library/zlib.html):

> (An Adler-32 checksum is almost as reliable as a CRC32 but can be computed much more quickly.)

Reply:
seems like fairly similar performance:

```
>>> byte_strings = [secrets.token_bytes(500) for i in range(1,1000000)]
>>> start = timeit.default_timer(); x = [zlib.adler32(s) for s in byte_strings]; end = timeit.default_timer()
>>> end-start
0.22351763100004973
>>> start = timeit.default_timer(); x = [zlib.crc32(s) for s in byte_strings]; end = timeit.default_timer()
>>> end-start
0.1954255269999976
```